File: Codeforces/1846+E2.cpp
#include <bits/stdc++.h>

using namespace std;

#define read(a, n) for(ll i = 0; i < n; ++i) cin >> a[i];
#define print(a, n) for(ll i = 0; i < n; ++i) if(i == n - 1){ cout << a[i] << endl; } else { cout << a[i] << ' '; }

typedef long long int ll;
typedef long double ld;

const ll mod = 1e9 + 7;
const ll mxn = 1e6 + 5;

void fast(){
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
}

bool isPerfectSquare(ll x)
{
    ll n = x;
    ll lo = 1;
    ll hi = x;

    while (hi - lo > 1) {
        ll mid = (hi + lo) / 2;
        if (x == 1) {
            return true;
        }
        else if (mid * mid == x) {
            return true;
        }
        else if (mid * mid > x) {
            hi = mid - 1;
        }
        else if (mid * mid < x) {
            lo = mid + 1;
        }
    }
    if (lo * lo == x || hi * hi == x) {
        return 1;
    }
    return 0;
}

ll floorSqrt(ll x){
    // Base cases
    if (x == 0 || x == 1){
        return x;
    }
 
    ll start = 1, end = x / 2, ans;
    while (start <= end) {
        ll mid = (start + end) / 2;
 
        ll sqr = mid * mid;
        if (sqr == x){
            return mid;
        }
        if (sqr <= x){
            start = mid + 1;
            ans = mid;
        }
        else{
            end = mid - 1;
        }
    }
    return ans;
}

ll fastPower(ll b, ll e){

    ll ans = 1;

    while(e){

        if(e & 1){
            ans *= b;
            --e;
        }
        else{
            b *= b;
            e >>= 1;
        }
    }

    return ans;
}

bool oneDistance(ll n){
    ll d = (4 * n) - 3;

    if(!isPerfectSquare(d)){
        return 0;
    }

    ll sqr = floorSqrt(d);
    --sqr;

    if(sqr % 2 != 0) return 0;
    if(sqr / 2 <= 1) return 0;
    return 1;
}

ll lim = 1e18 + 9;

int main(){
    // fast();

    set <ll> vc;
    ll l2 = 1000000;

    for(ll b = 2; b < l2; ++b){

        ll ans = 1 + b;
        ll c = b * b;

        for(ll n = 3; n < 64; ++n){

            ans += c;

            if(ans <= lim){
                vc.insert(ans);
                c *= b;
            }  
            else break;
        }
    }

    ll t;
    cin >> t;

    while(t--){
        ll n;
        cin >> n;

        if(oneDistance(n)){
            cout << "YES\n";
            continue;
        }

        if(vc.count(n)){
            cout << "YES\n";
        }
        else{
            cout << "NO\n";
        }
    }
    
    return 0;
}

File: Codeforces/1846+E2.py

def root(x):
    if x <= 1: return x

    st = 2
    ed = x // 2
    
    ans = -1

    while(st <= ed):
        mid = (st + ed) // 2

        sq = mid * mid

        if(sq == x): return mid
        elif(sq < x): 
            st = mid + 1
            ans = mid
        else:
            ed = mid - 1
    
    return ans

def isPerfectSquare(x):

    y = root(x)

    if(y * y == x):
        return True
    
    return False

def fastPower(b, e):

    ans = 1

    while(e >= 1):

        if(e % 2 == 1):
            ans *= b
            e -= 1

        b *= b
        e //= 2
    
    return ans

def oneDistance(n):
    d = (4 * n) - 3

    if(isPerfectSquare(d) == False):
       return 0
    
    sq = root(d)
    sq -= 1

    if (sq % 2 == 1): return 0
    if(sq / 2 <= 1): return 0

    return 1

def binarySearch(x, l):
    n = len(l)

    lo = 0
    hi = n - 1

    while(lo < hi):
        mid = (lo + hi) // 2

        if(l[mid] == x):
            return 1
        elif(l[mid] > x):
            hi = mid
        else:
            lo = mid + 1
    
    return 0


lim2 = 1000000
lim = 10000000000000000000

l = []

for b in range(2, lim2 + 1):

    for n in range(3, 64):
        ans = fastPower(b, n)

        if(ans <= lim):
            l.append(ans)


l.sort()

for i in range(0, 100):
    print(l[i], end = ' ')
print()


t = int(input())

while(t > 0):
    n = int(input())

    # x = (4 * n) - 3
    # print(root(x), root(x) * root(x), x)

    if(oneDistance(n)):
        print("YES")
    elif(binarySearch(n, l) == True):
        print("YES")
    else:
        print("NO")

    t -= 1


    
    